### ISSUE TITLE
Unsatisfiable requirements when changing compat versions of dependencies

### ISSUE THREAD

**Issue by mohamed82008** [JuliaLang/Pkg.jl#2303]
I found that I get a seemingly buggy and confusing unsatisfiable requirements error when the following happens. Say I have a package AlgebraicMultigrid.jl (AMG) whose latest version supports IterativeSolvers.jl versions [0.8.3, 0.9). I dev it to update the compat to 0.9 but it is now incompatible with 0.8.3. Note I didn't explicitly install IterativeSolvers, it was pulled as a dependency when installing AMG. So after I update the compat entry for IterativeSolvers in the AMG Project.toml file, I type `] resolve` and I get the following error:

```julia
ERROR: Unsatisfiable requirements detected for package IterativeSolvers [42fd0dbc]:
 IterativeSolvers [42fd0dbc] log:
 ├─possible versions are: [0.7.0-0.7.1, 0.8.0-0.8.5, 0.9.0] or uninstalled
 ├─restricted to versions 0.9 by AlgebraicMultigrid [2169fc97], leaving only versions 0.9.0
 │ └─AlgebraicMultigrid [2169fc97] log:
 │   ├─possible versions are: 0.4.0 or uninstalled
 │   └─AlgebraicMultigrid [2169fc97] is fixed to version 0.4.0
 └─restricted to versions 0.8.4 by an explicit requirement — no versions left
```

In particular, the buggy and confusing part is the line `restricted to versions 0.8.4 by an explicit requirement — no versions left` since I never told Pkg to "restrict" IterativeSolvers to 0.8.4. Simply typing `]add IterativeSolvers` installs the correct version, 0.9, and the above error doesn't show again when I resolve. I came across different forms of this error before many times but I could never come up with a minimal example to reproduce and report it. I hope this is helpful.

**Comment by KristofferC** [comment]
Bug reports about resolver issues pretty much need to also provide `Project.toml`, `Manifest.toml` and the exact Julia version to be actionable. Or at least steps to reproduce those.

**Comment by mohamed82008** [comment]
Steps to reproduce it in Julia 1.5.3:
1. `] dev AlgebraicMultigrid`
2. Change the compat entry for `IterativeSolvers` in AMG's Project.toml file from `IterativeSolvers = "0.8.3"` to `IterativeSolvers = "0.9"`
3. Run `] resolve`

**Comment by KristofferC** [comment]
Right now, `resolve` tries to keep the current versions fixed.

https://github.com/JuliaLang/Pkg.jl/blob/b2bc8627c2835e2b78cf7213e034f17130d534ce/src/API.jl#L331

You could use `update` instead.

**Comment by mohamed82008** [comment]
Hmm but I don't want to update any of the direct dependencies. Does it make sense to only update the indirect deps, or prompt the user to do so instead of throwing the above error? The above error correctly shows up in a different context when I pin the version of a package for example but in this case it's confusing because I didn't pin anything.

**Comment by dpinol** [comment]
I agree this is an issue. resolve should be able to upgrade dependencies when needed
```

Pkg.resolve(; io::IO=stderr)

  Update the current manifest with potential changes to the dependency graph from packages that are tracking a path
```
